packages/core/src/amazonqGumby/chat/storages/chatSession.ts

```diff
@@ -0,0 +1,38 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+import { Messenger } from '../controller/messenger/messenger'
+import { Session } from '../session/session'
+
+export class ChatSessionStorage {
+    private sessions: Map<string, Session> = new Map()
+
+    constructor(private readonly messenger: Messenger) {}
+
+    private async createSession(tabID: string): Promise<Session> {
+        const session = new Session(this.messenger, tabID)
+        this.sessions.set(tabID, session)
+        return session
+    }
+
+    public async getSession(tabID: string): Promise<Session> {
+        const currentSession = this.sessions.get(tabID)
+        if (currentSession === undefined) {
+            // If a session doesn't already exist just create it
+            return this.createSession(tabID)
+        }
+        return currentSession
+    }
+
+    // Find all sessions that are currently waiting to be authenticated
+    public getAuthenticatingSessions(): Session[] {
+        return Array.from(this.sessions.values()).filter(session => session.isAuthenticating)
```

Comment:
This entire file was copy-pasted from featureDev? https://github.com/aws/aws-toolkit-vscode/blob/46560e70b7dd29d83ff557b1db246db5091dd7fe/packages/core/src/amazonqFeatureDev/storages/chatSession.ts#L33

I don't see how it is sustainable that we keep copy-pasting modules everywhere.

Reply:
This file has gone under significant change, as a code transformation session does conceptually need to track different things than FeatureDev, which should alleviate your concern about copy/pasting code. Nothing in this file impacts global state in a way that should impact other features.

I agree that there is a better way to handle authentication -- there are patterns introduced in CodeWhisperer that were replicated by FeatureDev and now by Gumby, where everyone would benefit by having common classes to use. There's no base "chatTab" module that handles this sort of thing, thus the code duplication.

Right now, all 3 use a pattern where they poll for the current global state at specific points in the chat flow (when the chat opens, when a user interaction would cause an interaction with the backend, etc.) _If_ they see that the global state is 'unauthenticated',  all 3 show an error message to the user, update their local state to `isAuthenticating`, and then lock the tab. They then wait for an event to be fired by the global auth handler so that they can unlock the tab -- in the meantime, they require that local 'isAuthenticating' state to prevent certain user interactions.

Unfortunately I don't have much insight into this pattern beyond that ): I'd like to see if this entire pattern could be event-driven rather than partially so, but it would require introducing further complexity, and making changes to other teams' code -- this would be the sort of thing I'd want to design and get buy-in ahead of time for rather than trying to rush it in to launch a feature.

In the meantime, I will see if I can extract common behaviors to a utility module that can be shared between the apps!

Reply:
> I agree that there is a better way to handle authentication -- there are patterns introduced in CodeWhisperer that were replicated by FeatureDev and now by Gumby, where everyone would benefit by having common classes to use. 

Yes, exactly. 

> this would be the sort of thing I'd want to design and get buy-in ahead of time for rather than 

Ok, can you create a back log issue?

> trying to rush it in to launch a feature.

Sure, but we made similar compromises for re:Invent. We need to pay down this tech debt soon.

> I will see if I can extract common behaviors to a utility module that can be shared between the apps!

Thank you!